### ISSUE TITLE
`isUnmodifiable` assertion for Collections

### ISSUE THREAD

**Issue by scordio** [assertj/assertj-core#2102]
#### Summary
Considering the `java.util.Collections.unmodifiableXXX()` methods, we could add a `isUnmodifiable` method to the assert classes for collections.

#### Example

```java
List<Integer> list = Collections.unmodifiableList(Arrays.asList(1, 2, 3));
assertThat(list).isUnmodifiable();
```

**Comment by HaydenMeloche** [comment]
That would be cool! How would we assert the list is unmodifiable under the hood?
We could do:
```java
list.getClass().isInstance(Collections.unmodifiableList(Arrays.asList("any value")))
```
This should work if `UnmodifiableList` or `UnmodifiableRandomAccessList` is returned.

**Comment by joel-costigliola** [comment]
@HaydenMeloche that would cover the JDK types but if I use a Guava ImmutableCollection that won't work so the best implementation IMO is following the possible workaround by calling clear and catching the UnsupportedOperationException, if no exceptin is caught then fail the assertion

**Comment by HaydenMeloche** [comment]
@joel-costigliola Fair point. I'd be willing to tackle this if you think it's worth working on :)

**Comment by joel-costigliola** [comment]
Sure, go ahead, thanks!

**Comment by scordio** [comment]
@HaydenMeloche thank you! Before start working on it please let me put together a more detailed description of the approach I have in mind, so we can blame it also considering Guava. I should be able to do it today or tomorrow max.

**Comment by scordio** [comment]
I added some more details and I'll complete them during the weekend. However, this should give already a feeling so please let me know your thoughts.

I'd suggest starting the implementation of `AbstractCollectionAssert` which would inherit from `AbstractIterableAssert` and provide `isUnmodifiable`. The rest are extension points and could be done separately.

**Comment by HaydenMeloche** [comment]
@scordio Thanks for the detailed explanation. I'll get started when I have some time this week. I might create a draft PR when I have something concrete so I can get feedback

**Comment by HaydenMeloche** [comment]
hey @scordio got a chance to working on this over the weekend. I setup my assertion class as such
```java
public abstract class AbstractCollectionAssert <SELF extends AbstractCollectionAssert<SELF, ACTUAL, ELEMENT, ELEMENT_ASSERT>,
  ACTUAL extends java.util.Collection<? extends ELEMENT>,
  ELEMENT,
  ELEMENT_ASSERT extends AbstractAssert<ELEMENT_ASSERT, ELEMENT>>
  extends AbstractIterableAssert<SELF, ACTUAL, ELEMENT, ELEMENT_ASSERT>
  implements IndexedObjectEnumerableAssert<SELF, ELEMENT> {
```

I'm running into an issue now with this setup. Existing test cases that test the `Iterable` type are now looking for a Collection assert rather than a Iterator assert.

Ex: This test case:
https://github.com/assertj/assertj-core/blob/4fad9a03993e66fd4e2735352c22c52d206e9a1e/src/test/java/org/assertj/core/api/Assertions_assertThat_with_Iterable_Test.java#L30

Gives error:
```
Required type:
AbstractIterableAssert
<?,
java.lang.Iterable<? extends java.lang.Object>,
java.lang.Object,
org.assertj.core.api.ObjectAssert<java.lang.Object>>
Provided:
CollectionAssert
<Object,



no instance(s) of type variable(s) ELEMENT, T exist so that Collection<? extends ELEMENT> conforms to Iterable<?>
```
In [Assertions.java](https://github.com/assertj/assertj-core/blob/main/src/main/java/org/assertj/core/api/Assertions.java) I setup the collection assert as 
```java
  public static <ELEMENT> CollectionAssert<ELEMENT> assertThat(Collection<? extends ELEMENT> actual) {
    return AssertionsForInterfaceTypes.assertThat(actual);
  }
```

Since `Collection` extends `Iterable` I understand why this is happening just not sure what the recommended action would be... 
any ideas/help would be appreciated :) I can push the code up to a branch if it's easier.

**Comment by scordio** [comment]
Hi @HaydenMeloche, yes please, just push/raise a draft PR and I'll have a look.
